### ISSUE TITLE
cli++ proof of concept and document for MPC

### ISSUE THREAD

**Issue by katyashchukina** [uchicago-cs/chiventure#1185]
Currently, we lack a proof of concept on the  Micro Parsers Combinators library (https://github.com/orangeduck/mpc), a Parser Combinator library for C, to interact with command-line input. We need to show proof of concept for the  library. The process and findings will be documented under this issue.

asignees: kate, matteo
team: cli++

**Comment by mrestuccia1** [comment]
After exploring the Micro Parsers Combinators library, we found that MPC contains multiple proofs of concepts of its own, but two that initially stand out are the maths.c file and the token.c file. The maths.c returns all of the symbols in an equation separated, which will be useful, as we want to create a calculator as a proof of concept. The token.c seems very useful in general, as it tokenizes sentences into each different word which at first sight will likely be helpful to read and understand command line input.

**Comment by mrestuccia1** [comment]
I struggled to get the maths executable to work for some time, as it was unclear what the correct process to call it was. As such, I tried switching over to the token.c file, as I thought I could build a calculator using the string tokenization feature of the file. However, this process was blocked as it did not allow me to do loops to isolate the tokens into variables, because "loop initial declarations are only allowed in C99 or C11 mode". Thus, I went back to the maths.c file, and eventually figured out I needed to run it by passing in a file with an expression. This was ideal as the maths.c specifically recognizes the format of a math equation, understands what each component is, and maintains the order of operations. For example, for input file readme.maths containing the equation (4 * 2 + 3 / 7) - 6 it returns: 

[maths.c output.pdf](https://github.com/uchicago-cs/chiventure/files/8555312/maths.c.output.pdf).

I will now work on actually using these different pieces to create a calculator.  

NOTE: A possible problem with this way of calling the maths executable is that it needs a file to be inputted and it is not just enough to input something directly on the command line, though I'm sure this could be worked around if the command line input was redirected to a file which was then redirected to the maths.c call.

**Comment by mrestuccia1** [comment]
MPC lacks usage instructions and is therefore extremely hard to use. Some of the provided examples in the documentation provide syntax errors when run, so it's extremely difficult to know what is correct usage and what isn't. There are weird rules set in the library (such as: "error: ISO C90 forbids mixed declarations and code"). While these rules can be worked around, they are tedious to work around. Therefore, I do not believe that this is a viable option for our CLI needs. 

For this issue, Kate created the issue, and Matteo (me) carried out the attempts to create a proof of concept. Even though I tried multiple times, I was unable to come up with a good proof of concept due the difficulty of usage of MPC. I also documented everything in the process.
